server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,318 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.apache.lucene.search.TotalHits;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.query.QuerySearchResult;
+import org.elasticsearch.transport.RemoteClusterAware;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * Use this progress listener for cross-cluster searches where a single
+ * coordinator is used for all clusters (minimize_roundtrips=false).
+ * It updates state in the SearchResponse.Clusters object as the search
+ * progresses so that the metadata required for the _clusters/details
+ * section in the SearchResponse is accurate.
+ */
+public class CCSSingleCoordinatorSearchProgressListener extends SearchProgressListener {
+
+    private SearchResponse.Clusters clusters;
+    private TransportSearchAction.SearchTimeProvider timeProvider;
+
+    /**
+     * Executed when shards are ready to be queried (after can-match)
+     *
+     * @param shards The list of shards to query.
+     * @param skipped The list of skipped shards.
+     * @param clusters The statistics for remote clusters included in the search.
+     * @param fetchPhase <code>true</code> if the search needs a fetch phase, <code>false</code> otherwise.
+     **/
+    @Override
+    public void onListShards(
+        List<SearchShard> shards,
+        List<SearchShard> skipped,
+        SearchResponse.Clusters clusters,
+        boolean fetchPhase,
+        TransportSearchAction.SearchTimeProvider timeProvider
+    ) {
+        assert clusters.isCcsMinimizeRoundtrips() == false : "minimize_roundtrips must be false to use this SearchListener";
+
+        this.clusters = clusters;
+        this.timeProvider = timeProvider;
+
+        // Partition by clusterAlias and get counts
+        Map<String, Integer> skippedByClusterAlias = partitionCountsByClusterAlias((skipped));
+        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias((shards));
+        // the 'shards' list does NOT include the shards in the 'skipped' list, so to get total counts, combine counts from both
+        skippedByClusterAlias.forEach((cluster, skipCount) -> totalByClusterAlias.merge(cluster, skipCount, Integer::sum));
+
+        for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
+            String clusterAlias = entry.getKey();
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
```

Comment:
Not totally related to this PR just a comment :) but we could probably just use a CHM in `Clusters`? That seems like it would save some complexity down the line as well in serialisation etc.

Reply:
I'll create an GH issue for that change.

Reply:
Issue created: https://github.com/elastic/elasticsearch/issues/99101